File: Difficulty: Medium/Diameter of a Binary Tree/diameter-of-a-binary-tree.cpp
/*
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};

Node* newNode(int val) {
    return new Node(val);
}
*/

class Solution {
private:

    int max_height(Node* root, int& res){
        if(!root) return 0;
        
        //if node has left tree
        int leftHeight=max_height(root->left, res);
        int rightHeight=max_height(root->right,res);

        res = max(res,leftHeight+rightHeight);
        
        return max(leftHeight,rightHeight)+ 1;
    }
    
  public:
    int diameter(Node* root) {
        // Your code here
        // ) return 0;
        
        int res=0;
        max_height(root,res);
        return res;
    }
};

File: Difficulty: Medium/Diameter of a Binary Tree/diameter-of-a-binary-tree.py
'''
# Tree Node
class Node:
    def __init__(self, val):
        self.data = val
        self.right = None
        self.left = None
'''

class Solution:
    
    def solve(self, root):
        if root is None:
            return [0,0]
        
        # if it is a leaf node
        if root.left is None and root.right is None:
            return [0,1]
        
        left_h = self.solve(root.left)
        right_h = self.solve(root.right)
        
        # Get the previous max Diamter 
        maxD = max(left_h[0],right_h[0])
        
        # Compute the max Diamter till now with the diamter at the current node
        maxD = max(maxD, left_h[1] + right_h[1])
        
        # Caculate the max abs height on left / right on the node 
        h = 1 + max(left_h[1], right_h[1])
        
        return [maxD, h]
    
    def diameter(self, root):
        # code here
        # pair : maxDiameter, height
        max_height_pair = self.solve(root)
        return max_height_pair[0]
        